File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        l = 0
        r = 0
        count = 0
        myDict = {}
        while(r < len(s)):
            myDict[s[r]] = myDict.get(s[r], 0) + 1
            while len(myDict) == 3:
                count+= (len(s) - r)
                if myDict[s[l]] == 1:
                    del myDict[s[l]]
                else:
                    myDict[s[l]]-=1
                l+=1
            r+=1
        return count

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.java
class Solution {
    public int numberOfSubstrings(String s) {
        int left = 0;
        int right = 0;
        int [] index = new int[3];
        Arrays.fill(index,-1);
        int res = 0;
        while(left < s.length()){
            while((right < s.length()) && (index[0] == -1 || index[1] == -1 || index[2] == -1)){
                index[s.charAt(right) - 'a'] = right;
                right+=1;
            }
            if(right == s.length() && ((index[0] == -1 || index[1] == -1 || index[2] == -1))){
                break;
            }
            res += (s.length() - right + 1);
            if(index[s.charAt(left) - 'a'] == left){
                index[s.charAt(left) - 'a'] = -1;
            }
            left +=1;
        }
        return res;
    }
}